modules/fedimint-mint-client/src/backup/tracker.rs

```diff
@@ -0,0 +1,716 @@
+        let mut s = Self {
+            start_epoch: backup.epoch_count,
+            next_epoch: backup.epoch_count,
+            end_epoch: current_epoch,
+            spendable_note_by_nonce: backup
+                .notes
+                .into_iter()
+                .map(|(amount, note)| (note.note.0, (amount, note)))
+                .collect(),
+            pending_outputs: backup
+                .pending_notes
+                .into_iter()
+                .map(|(out_point, issuance_requests)| {
+                    (
+                        out_point,
+                        (
+                            issuance_requests
+                                .notes
+                                .iter_items()
+                                .map(|(amount, iss_req)| {
+                                    (amount, (iss_req.recover_blind_nonce().0, Some(*iss_req)))
+                                })
+                                .collect(),
+                            HashMap::default(),
+                        ),
+                    )
+                })
+                .collect(),
+            pending_nonces: HashMap::default(),
+            next_pending_note_idx: backup.next_note_idx.clone(),
+            last_mined_nonce_idx: backup.next_note_idx,
+            threshold: pub_key_shares.threshold(),
+            gap_limit,
+            tbs_pks,
+            pub_key_shares,
+        };
+
+        for amount in amount_tiers {
+            s.fill_initial_pending_nonces(amount);
+        }
+
+        s
+    }
+
+    /// Fill each tier pool to the gap limit
+    fn fill_initial_pending_nonces(&mut self, amount: Amount) {
+        info!(%amount, count=self.gap_limit, "Generating initial set of nonces for amount tier");
+        for _ in 0..self.gap_limit {
+            self.add_next_pending_nonce_in_pending_pool(amount);
+        }
+    }
+
+    /// Add next nonce from `amount` tier to the `next_pending_note_idx`
+    fn add_next_pending_nonce_in_pending_pool(&mut self, amount: Amount) {
+        let note_idx_ref = self.next_pending_note_idx.get_mut_or_default(amount);
+
+        let (note_issuance_request, blind_nonce) = NoteIssuanceRequest::new(
+            secp256k1::SECP256K1,
+            MintClientModule::new_note_secret_static(&self.secret, amount, *note_idx_ref),
+        );
+        assert!(self
+            .pending_nonces
+            .insert(
+                blind_nonce.0,
+                (note_issuance_request, *note_idx_ref, amount)
+            )
+            .is_none());
+
+        note_idx_ref.advance();
+    }
+
+    pub fn handle_input(&mut self, input: &MintInput) {
+        // We attempt to delete any nonce we see as spent, simple
+        for (_amt, note) in input.0.iter_items() {
+            self.spendable_note_by_nonce.remove(&note.0);
+        }
+    }
+
+    pub fn handle_output(&mut self, out_point: OutPoint, output: &MintOutput) {
```

Comment:
A lot of this logic is copied over, right? Anything I should pay more attention on?